### PR TITLE
Feature Level Visibility tutorial

### PR DESCRIPTION
**Pull request by jdk2pq**

### CODE REVIEW

File: _posts/2015-03-30-geomesa-feature-level-visibility.md [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+
+Additionally, if you look at line 169, you'll see:
+
+{% highlight scala %}
+if (i % 2 == 0) {
+    simpleFeature.setAttribute("Visibility", "admin");
+    simpleFeature.getUserData().put(SecurityUtils.FEATURE_VISIBILITY, "admin")
+} else {
+    simpleFeature.setAttribute("Visibility", "user|admin");
+    simpleFeature.getUserData().put(SecurityUtils.FEATURE_VISIBILITY, "user|admin");
```
Comment (elahrvivaz) [inline _posts/2015-03-30-geomesa-feature-level-visibility.md]: might want to change these methods to use the static calls, which are a bit easier: org.locationtech.geomesa.core.security.SecurityUtils#setFeatureVisibility

File: _posts/2015-03-30-geomesa-feature-level-visibility.md [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+<style>
+code.xml { color:#93a1a1 }
+</style>
+
+## Introduction
+
+Feature visibility is a new and important feature recently supported by GeoMesa. Accumulo has support for cell-level visibility built-in, and in recent versions of GeoMesa,
+support was added to utilize these visibility markings and enhance data security.
+<!--more-->
+From the Accumulo userguide:
```
Comment (elahrvivaz) [inline _posts/2015-03-30-geomesa-feature-level-visibility.md]: user guide

File: _posts/2015-03-30-geomesa-feature-level-visibility.md [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+
+This means that the feature `simpleFeature` will be written to GeoMesa with the visibility `admin`, where only users that have been granted the authorization level of `admin` will be able to view the feature in GeoMesa.
+
+That's it for changes in the Quickstart project! Easy, right? So, to add visibility to a GeoMesa feature, simply add a line that looks like this when creating or writing a SimpleFeature:
+
+{% highlight scala %}
+simpleFeature.getUserData().put(SecurityUtils.FEATURE_VISIBILITY, "{visibility-string-here}");
+{% endhighlight %}
+
+Now, when users query the GeoMesa, or when data are read out of Accumulo by GeoServer, only users with the correct level of authorization will be able to read the feature.
```
Comment (elahrvivaz) [inline _posts/2015-03-30-geomesa-feature-level-visibility.md]: the geomesa

Comment (elahrvivaz) [inline _posts/2015-03-30-geomesa-feature-level-visibility.md]: data are? not sure where we fall on this.

File: _posts/2015-03-30-geomesa-feature-level-visibility.md [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+
+> Accumulo extends the BigTable data model to implement a security mechanism known as cell-level security. Every key-value pair has its own security label, stored under the column visibility element of the key, which is used to determine whether a given user meets the security requirements to read the value. This enables data of various security levels to be stored within the same row, and users of varying degrees of access to query the same table, while preserving data confidentiality.
+
+In this tutorial, you'll be guided through ingesting data with varying levels of visibility and querying that data as different users through GeoServer.
+
+## Prerequistes
+
+If you haven't already [read through the Quickstart tutorial](http://www.geomesa.org/2014/05/28/geomesa-quickstart/) and
+gone through the initial setup of GeoMesa, read through that first. We'll be using a customized version of the data generated by the GeoMesa Quickstart project.
+
```
Comment (elahrvivaz) [inline _posts/2015-03-30-geomesa-feature-level-visibility.md]: Add something about 'user should understand accumulo visibilities' and link to the accumulo user guide on them.

File: _posts/2015-03-30-geomesa-feature-level-visibility.md [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+In the Quickstart example, 500 features are written with the visibility `admin` and 500 are written with `user|admin`. However, unless users are explicitly granted permissions to read these tables and read features with that level of authorization, we cannot visualize the data in a meaningful way. Let's continue on and add new Accumulo users to do just that.
+
+## Adding new Accumulo users
+
+In this part of the tutorial, we'll get familiar with the Accumulo shell. First, login to the Accumulo shell with:
+
+{% highlight bash %}
+accumulo shell -u {username-here} -p {password-here}
+# OR
```
Comment (elahrvivaz) [inline _posts/2015-03-30-geomesa-feature-level-visibility.md]: Does this OR show up as a title, or is that just github?

File: _posts/2015-03-30-geomesa-feature-level-visibility.md [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+
+Download this XML file, or copy the contents, and add it as a Style in GeoServer. It will style the points on a map based on the visibility attribute present.
+
+Lastly, click on "Layer Preview" in the left hand sidebar and find your two newly added layers. If everything went correctly, you should see fewer results returning in the `user` layer than in the `admin` layer, and this is expected behavior. Because `user` has only been granted permission to view features with the `user` visibility, only those 500 records are returned. However, the `admin` user is granted permission to see both `admin` and `user` visibilities.
+
+## Expanding this concept
+
+In this very simple example, you wrote features of two different visibilities, added two new users to Accumulo, and granted them separate authorization levels to be able to view portions of the data.
+This tutorial has real-world use cases in security and data integrity. For example, when storing data at differing classification levels in the same database, and having users at varying levels of security,
```
Comment (elahrvivaz) [inline _posts/2015-03-30-geomesa-feature-level-visibility.md]: not sure if we should mention classification here.

File: _posts/2015-03-30-geomesa-feature-level-visibility.md [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+The concept of feature level visibility can be extended and modified to have many more, or only a few, visibility levels. And with GeoServer being flexible and extensible, writing a module to consider feature
+level security in GeoServer is relatively painless.
+
+At CCRi, we've implemented a very small, example Java class that does a simple match on the username provided by GeoServer. Our example class then
+matches the username to a level of authorization specified in the class and hands this off to an AuthorizationsProvider in GeoMesa. This AuthorizationsProvider uses
+the correct level of authorization when querying Accumulo, no matter the user that is specified in the GeoServer DataStore. This allows us to have only one layer per
+DataSource, rather than what is in the example above, that many users of varying authorizations levels can hit and return only the data granted to their level of security.
+
+If you're interested in learning more about this, feel free to email [jdk2pq@ccri.com](mailto:jdk2pq@ccri.com).
```
Comment (elahrvivaz) [inline _posts/2015-03-30-geomesa-feature-level-visibility.md]: Should probably put the geomesa-users or dev list here instead of your personal email.

Comment (elahrvivaz) [inline _posts/2015-03-30-geomesa-feature-level-visibility.md]: or at least geomesa@ccri if we want to keep it off public lists.

**Comment by elahrvivaz** [comment]
Looks good. Could we add a link under 'expanding this concept' to the par of the 'authorizations' tutorial we had up already that shows how to configure ldap for different users? Would be useful here too I think.

File: _posts/2015-03-30-geomesa-feature-level-visibility.md (hunk 5 above)
Comment (jdk2pq) [inline _posts/2015-03-30-geomesa-feature-level-visibility.md]: Just GitHub.

**Comment by jdk2pq** [comment]
Committed as 9892eab6c1a878d0dfdfe951ed21a3ade94e5697

**Pull request closed without merging**